File: vns-python/pyvns/processor.py
import copy
from typing import Any, Final, NoReturn

from ._exception import ScriptCompilerException
from .content import Content
from .naming import Naming


# 视觉小说脚本编译器
class Processor:
    # 文件格式后缀
    SCRIPTS_FILE_EXTENSION: Final[str] = ".vns"

    __ALTER: dict[str, str] = {
        "[lang]": "[language]",
        "[opt]": "[option]",
        "[disp]": "[display]",
    }

    def __init__(self) -> None:
        self.__path_in: str = ""
        self.__line_index: int = 0
        self.__output: dict[str, dict[str, dict[str, Any]]] = {}
        self.__current_data: Content = Content({}, "head")
        self.__id: int = -1
        self.__lang: str = ""
        self.__section: str = ""
        self.__previous: str | None = None
        self.__lines: list[str] = []
        self.__branch_labels: dict[str, str] = {}
        self.__dialog_associate_key: dict[int, str] = {}
        self.__accumulated_comments: list[str] = []
        self.__blocked: bool = False

    # 获取id
    def get_id(self) -> int:
        return self.__id

    # 获取语言
    def get_language(self) -> str:
        return self.__lang

    # 转换一个str
    @staticmethod
    def __ensure_not_null(text: str) -> str | None:
        return None if text.lower() == "null" or text.lower() == "none" else text

    # 将参数分离出来
    @classmethod
    def __extract_parameter(cls, text: str, prefix: str) -> str | None:
        return cls.__ensure_not_null(cls.__extract_string(text, prefix))

    # 将字符串内容分离出来
    @staticmethod
    def __extract_string(text: str, prefix: str) -> str:
        sharp_index: int = text.find("#")
        return (
            text.removeprefix(prefix)
            if sharp_index < 0
            else text[:sharp_index].removeprefix(prefix)
        )

    # 编译失败
    def __terminated(self, _reason: str) -> NoReturn:
        raise ScriptCompilerException(
            f'File "{self.__path_in}", line {self.__line_index + 1}\n  {self.__get_current_line()}\nFail to compile due to {_reason}'
            if self.__line_index < len(self.__lines)
            else f'File "{self.__path_in}", Fail to compile due to {_reason}'
        )

    # 获取当前行
    def __get_current_line(self) -> str:
        return self.__lines[self.__line_index].lstrip()

    # 获取输出
    def get_output(self) -> dict[str, dict[str, dict[str, Any]]]:
        return self.__output

    # 处理数据
    def process(self, _path: str) -> None:
        self.__path_in = _path
        current_index: int = 0
        if self.__path_in.endswith(self.SCRIPTS_FILE_EXTENSION):
            with open(self.__path_in, "r", encoding="utf-8") as f:
                self.__lines = f.readlines()
        # 如果文件为空
        if len(self.__lines) <= 0:
            self.__terminated("Cannot convert an empty script file!")
        else:
            last_label: str | None = None
            # 预处理文件
            for index in range(len(self.__lines)):
                self.__lines[index] = self.__lines[index].removesuffix("\n")
                if self.__lines[index].startswith("[label]"):
                    if last_label is not None:
                        self.__terminated("The label is overwriting the previous one")
                    last_label = self.__extract_parameter(
                        self.__lines[index], "[label]"
                    )
                if self.__lines[index].startswith("[section]"):
                    current_index = 0
                elif (
                    not self.__lines[index].startswith("[")
                    and ":" in self.__lines[index]
                ):
                    self.__dialog_associate_key[index] = (
                        f"${current_index}" if current_index != 0 else "head"
                    )
                    current_index += 1
                    # 将id与label关联
                    if last_label is not None:
                        self.__branch_labels[last_label] = self.__dialog_associate_key[
                            index
                        ]
                        last_label = None
            if last_label is not None:
                print(f"The last label call {last_label} is not necessary!")
        self.__convert(0)
        self.__lines.clear()
        # 确保重要参数已被初始化
        if self.__id < 0:
            ScriptCompilerException("You have to set a nonnegative id!")
        elif len(self.__lang) <= 0:
            ScriptCompilerException("You have to set lang!")
        elif len(self.__section) <= 0:
            ScriptCompilerException("You have to set section!")

    # 转换
    def __convert(self, staring_index: int) -> None:
        self.__line_index = staring_index
        while self.__line_index < len(self.__lines):
            _currentLine: str = self.__get_current_line()
            if len(_currentLine) <= 0 or self.__lines[self.__line_index].startswith(
                "#"
            ):
                pass
            elif _currentLine.startswith("//"):
                self.__accumulated_comments.append(_currentLine.lstrip("//").lstrip())
            elif _currentLine.startswith("["):
                _tag: str = _currentLine[: _currentLine.index("]") + 1]
                match self.__ALTER.get(_tag, _tag):
                    # 背景图片
                    case "[bgi]":
                        self.__current_data.background_image = self.__extract_parameter(
                            _currentLine, _tag
                        )
                    # 背景音乐
                    case "[bgm]":
                        self.__current_data.background_music = self.__extract_parameter(
                            _currentLine, _tag
                        )
                    # 角色进场
                    case "[show]":
                        for _name in self.__extract_string(_currentLine, _tag).split():
                            self.__current_data.character_images.append(_name)
                    # 角色退场
                    case "[hide]":
                        for _name in self.__extract_string(_currentLine, _tag).split():
                            # 清空角色列表
                            if _name == "*":
                                self.__current_data.character_images.clear()
                                break
                            # 移除角色
                            for i in range(len(self.__current_data.character_images)):
                                if Naming(
                                    self.__current_data.character_images[i]
                                ).equal(_name):
                                    self.__current_data.character_images.pop(i)
                                    break
                    # 清空角色列表，然后让角色重新进场
                    case "[display]":
                        self.__current_data.character_images.clear()
                        for _name in self.__extract_string(_currentLine, _tag).split():
                            self.__current_data.character_images.append(_name)
                    # 章节id
                    case "[id]":
                        _id: str | None = self.__extract_parameter(_currentLine, _tag)
                        if _id is not None:
                            self.__id = int(_id)
                        else:
                            self.__terminated("Chapter id cannot be None!")
                    # 语言
                    case "[language]":
                        self.__lang = self.__extract_string(_currentLine, _tag)
                    # 部分
                    case "[section]":
                        if self.__previous is not None:
                            self.__output[self.__section][self.__previous][
                                "next"
                            ] = None
                        self.__section = self.__extract_string(_currentLine, _tag)
                        self.__output[self.__section] = {}
                        self.__output[self.__section]["head"] = {}
                        self.__current_data = Content({}, "head")
                        self.__previous = None
                    # 结束符
                    case "[end]":
                        assert self.__previous is not None
                        self.__output[self.__section][self.__previous]["next"] = None
                        break
                    # 转换场景
                    case "[scene]":
                        assert self.__previous is not None
                        self.__output[self.__section][self.__previous]["next"][
                            "type"
                        ] = "scene"
                        self.__current_data.background_image = self.__extract_parameter(
                            _currentLine, _tag
                        )
                        self.__blocked = True
                    # 终端
                    case "[block]":
                        if self.__previous is not None:
                            self.__output[self.__section][self.__previous][
                                "next"
                            ] = None
                        self.__current_data = Content({}, "id_needed")
                        self.__previous = None
                    # 选项
                    case "[option]":
                        # 确认在接下来的一行有branch的label
                        if not self.__lines[self.__line_index + 1].startswith("[br]"):
                            self.__terminated(
                                f"For option on line {self.__line_index + 1}, a branch label is not found on the following line"
                            )
                        # 如果next没被初始化，则初始化
                        assert self.__previous is not None
                        if (
                            self.__output[self.__section][self.__previous].get("next")
                            is None
                        ):
                            self.__output[self.__section][self.__previous]["next"] = {}
                        # 获取对应的下一个对话字典的指针
                        dialog_next: dict[
                            str, str | list[dict[str, str]]
                        ] = self.__output[self.__section][self.__previous]["next"]
                        if dialog_next.get("type") != "option":
                            dialog_next["type"] = "option"
                            dialog_next["target"] = []
                        assert isinstance(dialog_next["target"], list)
                        dialog_next["target"].append(
                            {
                                "text": self.__extract_string(_currentLine, "[option]"),
                                "id": self.__branch_labels[
                                    self.__extract_string(
                                        self.__lines[self.__line_index + 1], "[br]"
                                    )
                                ],
                            }
                        )
                        self.__line_index += 1
                    case _:
                        self.__terminated(f"invalid tag {_tag}")
            elif ":" in _currentLine:
                _narrator: str | None = self.__ensure_not_null(
                    _currentLine.removesuffix(" ").removesuffix(":")
                )
                self.__current_data.narrator = (
                    _narrator if _narrator is not None else ""
                )
                # 获取讲述人可能的立绘名称
                narrator_possible_images: tuple[str, ...] = tuple()
                if self.__current_data.narrator.lower() in Naming.get_database():
                    narrator_possible_images = tuple(
                        Naming.get_database()[self.__current_data.narrator.lower()]
                    )
                # 检查名称列表，更新character_images以确保不在说话的人处于黑暗状态
                for i in range(len(self.__current_data.character_images)):
                    _name_data: Naming = Naming(self.__current_data.character_images[i])
                    if _name_data.name in narrator_possible_images:
                        if "silent" in _name_data.tags:
                            _name_data.tags.remove("silent")
                    else:
                        _name_data.tags.add("silent")
                    self.__current_data.character_images[i] = _name_data.to_string()
                # 更新对话内容
                self.__current_data.contents.clear()
                for sub_index in range(self.__line_index + 1, len(self.__lines)):
                    if self.__lines[sub_index].startswith("- "):
                        self.__current_data.contents.append(
                            self.__extract_string(self.__lines[sub_index], "- ")
                        )
                    else:
                        break
                # 确认section不为None，如果为None，则警告
                if len(self.__section) <= 0:
                    self.__terminated("You have to specify section before script")
                # 如果section未在字典中，则初始化对应section的数据
                elif self.__section not in self.__output:
                    self.__output[self.__section] = {}
                # 如果上个dialog存在（不一定非得能返回）
                if self.__previous is not None:
                    if not self.__blocked:
                        self.__current_data.previous = self.__previous
                    else:
                        self.__current_data.previous = None
                        self.__blocked = False
                    # 生成数据
                    last_ref: dict[str, Any] | None = self.__output[self.__section].get(
                        self.__previous
                    )
                    if last_ref is not None:
                        if last_ref.get("next") is not None:
                            last_ref["next"]["target"] = self.__dialog_associate_key[
                                self.__line_index
                            ]
                            if "type" not in last_ref["next"]:
                                last_ref["next"]["type"] = "default"
                        else:
                            last_ref["next"] = {
                                "target": self.__dialog_associate_key[
                                    self.__line_index
                                ],
                                "type": "default",
                            }
                    else:
                        self.__terminated(f"KeyError: {self.__previous}")
                else:
                    self.__current_data.previous = None
                # 添加注释
                if len(self.__accumulated_comments) > 0:
                    self.__current_data.comments = self.__accumulated_comments
                    self.__accumulated_comments = []
                # 更新key
                self.__previous = self.__dialog_associate_key[self.__line_index]
                # 更新缓存参数
                self.__line_index += len(self.__current_data.contents)
                self.__output[self.__section][self.__previous] = copy.deepcopy(
                    self.__current_data.to_dict()
                )
                # 移除注释
                self.__current_data.comments.clear()
            else:
                self.__terminated("Invalid code or content!")
            self.__line_index += 1

File: vns-cpp/processor.cpp
#include <string>
#include <vector>
#include "processor.h"
#include <stdexcept>
#include "functions.h"


int Processor::get_id() const
{
	return id_;
}

std::string Processor::get_language() const
{
	return lang_;
}

std::string Processor::ensure_not_null(const std::string& text)
{
	return (iequals(text, "null") || iequals(text, "none")) ? "" : text;
}

std::string Processor::extract_parameter(const std::string& text)
{
	return ensure_not_null(extract_string(text));
}

std::string Processor::extract_tag(const std::string& text)
{
	return text.substr(text.find(TAG_STARTS) + 1, text.find(TAG_ENDS));
}

std::string Processor::extract_string(const std::string& text)
{
	return trim(text.substr(text.find(TAG_ENDS) + 1));
}

void Processor::terminated(const std::string& reason) const
{
	throw std::runtime_error("File \"" + path_in_ + "\", line " + std::to_string(line_index_ + 1) +
		"\n  " + get_current_line() + "\nFail to compile due to " + reason);
}

std::string Processor::get_current_line() const
{
	return lines_[line_index_];
}

std::unordered_map<std::string, std::unordered_map<std::string, std::unordered_map<std::string, std::any>>>
Processor::get_output() const
{
	return output_;
}

/*
void Processor::process(const std::string& path) {
    path_in_ = path;
    int current_index = 0;

    if (path_in_.ends_with(SCRIPTS_FILE_EXTENSION)) {
        std::ifstream file(path_in_);
        std::string line;
        while (std::getline(file, line)) {
            lines_.emplace_back(line);
        }
    }

    if (lines_.empty()) {
        terminated("Cannot convert an empty script file!");
    }

    std::string last_label;

    for (int index = 0; index < lines_.size(); ++index) {
        lines_[index] = trim(lines_[index].substr(0, lines_[index].find("#")));

        if (lines_[index].starts_with(TAG_STARTS)) {
            if (auto tag = extract_tag(lines_[index]); tag == "label") {
                if (!last_label.empty()) {
                    terminated("This label is overwriting the previous one");
                }

                last_label = extract_parameter(lines_[index]);
                if (RESERVED_WORDS.contains(last_label)) {
                    terminated("You cannot use reserved word '" + last_label + "' as a label");
                }
            }
            else if (tag == "section") {
                current_index = 0;
            }
        }
        else if (lines_[index].find(":") != std::string::npos) {
            dialog_associate_key_[index] = (current_index == 0 ? "head" : (current_index < 10 ? "~0" + std::to_string(current_index) : "~" + std::to_string(current_index)));

            last_label = "";
            ++current_index;
        }
    }

    convert(0);
    lines_.clear();

    if (id_ < 0) {
        terminated("You have to set a nonnegative id!");
    }
    else if (lang_.empty()) {
        terminated("You have to set lang!");
    }
    else if (section_.empty()) {
        terminated("You have to set section!");
    }
}

void Processor::convert(int starting_index) {
    line_index_ = starting_index;

    while (line_index_ < lines_.size()) {
        std::string currentLine = get_current_line();

        if (currentLine.empty() || lines_[line_index_].find("#") == 0) {
            // Skip empty lines or lines starting with #
            // Do nothing
        }
        else if (currentLine.find("//") == 0) {
            // Accumulate comments
            accumulated_comments_.push_back(currentLine.substr(2));
        }
        else if (currentLine.find(TAG_STARTS) == 0) {
            std::string tag = extract_tag(currentLine);

            switch (ALTERNATIVES.at(tag)) {
            case "bgi": {
                current_data_.background_image_ = extract_parameter(currentLine);
                break;
            }
            case "bgm": {
                current_data_.background_music_ = extract_parameter(currentLine);
                break;
            }
            case "show": {
                auto names = extract_string(currentLine);
                std::istringstream iss(names);
                std::copy(std::istream_iterator<std::string>(iss),
                    std::istream_iterator<std::string>(),
                    std::back_inserter(current_data_.character_images_));
                break;
            }
            case "hide": {
                auto names = extract_string(currentLine);
                if (names == "*") {
                    current_data_.character_images_.clear();
                }
                else {
                    current_data_.character_images_.erase(
                        std::remove_if(current_data_.character_images_.begin(),
                            current_data_.character_images_.end(),
                            [&](const std::string& name) {
                                return Naming(name).Equal(names);
                            }),
                        current_data_.character_images_.end());
                }
                break;
            }
            case "display": {
                current_data_.character_images_.clear();
                auto names = extract_string(currentLine);
                std::istringstream iss(names);
                std::copy(std::istream_iterator<std::string>(iss),
                    std::istream_iterator<std::string>(),
                    std::back_inserter(current_data_.character_images_));
                break;
            }
            case "id": {
                std::string id_str = extract_parameter(currentLine);
                if (!id_str.empty()) {
                    id_ = std::stoi(id_str);
                }
                else {
                    terminated("Chapter id cannot be None!");
                }
                break;
            }
            case "language": {
                lang_ = extract_string(currentLine);
                break;
            }
            case "section": {
                if (!previous_.empty()) {
                    output_[section_][previous_]["next"] = nullptr;
                }
                section_ = extract_string(currentLine);
                output_[section_] = {};
                output_[section_]["head"] = {};
                current_data_ = Content({}, "head");
                previous_ = "";
                break;
            }
            case "end": {
                assert(!previous_.empty());
                output_[section_][previous_]["next"] = nullptr;
                break;
            }
            case "scene": {
                assert(!previous_.empty());
                output_[section_][previous_]["next"]["type"] = "scene";
                current_data_.background_image_ = extract_parameter(currentLine);

                if (!current_data_.background_image_.empty() && current_data_.background_image_.length() == 0) {
                    current_data_.background_image_ = nullptr;
                }
                blocked_ = true;
                break;
            }
            case "block": {
                if (!previous_.empty()) {
                    output_[section_][previous_]["next"] = nullptr;
                }
                current_data_ = Content({}, "id_needed");
                previous_ = "";
                break;
            }
            case "option": {
                if (current_data_.contents_.empty()) {
                    terminated("Invalid option syntax: '->' cannot be found!");
                }

                assert(!previous_.empty());
                if (output_[section_][previous_]["next"] == nullptr) {
                    output_[section_][previous_]["next"] = {};
                }

                auto& next_ref = output_[section_][previous_]["next"];
                if (next_ref["type"] != "options") {
                    next_ref["type"] = "options";
                    next_ref["target"] = std::vector<std::unordered_map<std::string, std::variant<std::string, int>>>();
                }

                auto src_to_target = extract_string(currentLine);
                auto& target_list = std::get<std::vector<std::unordered_map<std::string, std::variant<std::string, int>>>>(next_ref["target"]);

                target_list.push_back({
                    {"text", src_to_target.substr(0, src_to_target.find("->")).strip()},
                    {"id", ensure_not_null(src_to_target.substr(src_to_target.find("->") + 2).strip())}
                    });

                break;
            }
            case "label": {
                // Placeholder, no action needed for "label" tag in C++ implementation
                break;
            }
            default: {
                terminated("Invalid tag " + tag);
                break;
            }
            }
        }
        else if (currentLine.find(":") != std::string::npos) {
            std::string narrator = ensure_not_null(currentLine.substr(0, currentLine.size() - 1));
            current_data_.narrator_ = narrator;

            // Rest of the logic for processing dialog content
        narrator_possible_images: tuple[str, ...] = tuple()
            if (current_data_.narrator_.lower() in Naming.get_database()) :
                narrator_possible_images = tuple(
                    Naming.get_database()[current_data_.narrator_.lower()]
                )
                for (int i = 0; i < current_data_.character_images_.size(); ++i) {
                    Naming name_data(current_data_.character_images_[i]);
                    if (std::find(narrator_possible_images.begin(), narrator_possible_images.end(), name_data.GetName()) != narrator_possible_images.end()) {
                        if (name_data.GetTags().count("silent")) {
                            name_data.GetTags().erase("silent");
                        }
                    }
                    else {
                        name_data.GetTags().insert("silent");
                    }
                    current_data_.character_images_[i] = name_data.ToString();
                }

        current_data_.contents_.clear();

        for (int sub_index = line_index_ + 1; sub_index < lines_.size(); ++sub_index) {
            if (lines_[sub_index].find("- ") == 0) {
                current_data_.contents_.push_back(lines_[sub_index].substr(2).strip());
            }
            else {
                break;
            }
        }

        if (section_.empty()) {
            terminated("You have to specify section before script");
        }
        else if (output_.find(section_) == output_.end()) {
            output_[section_] = {};
        }

        if (!previous_.empty()) {
            if (!blocked_) {
                current_data_.previous_ = previous_;
            }
            else {
                current_data_.previous_ = nullptr;
                blocked_ = false;
            }

            auto last_ref = output_[section_].find(previous_);
            if (last_ref != output_[section_].end()) {
                if (last_ref->second.find("next") != last_ref->second.end()) {
                    auto& next_ref = last_ref->second["next"];
                    if (next_ref.find("type") != next_ref.end() && next_ref["type"] != "options") {
                        next_ref["target"] = dialog_associate_key_[line_index_];
                        if (next_ref.find("type") == next_ref.end()) {
                            next_ref["type"] = "default";
                        }
                    }
                    else {
                        next_ref = {
                            {"target", dialog_associate_key_[line_index_]},
                            {"type", "default"}
                        };
                    }
                }
                else {
                    last_ref->second["next"] = {
                        {"target", dialog_associate_key_[line_index_]},
                        {"type", "default"}
                    };
                }
            }
            else {
                terminated("KeyError: " + previous_);
            }
        }
        else {
            current_data_.previous_ = nullptr;
        }

        if (!accumulated_comments_.empty()) {
            current_data_.comments_ = accumulated_comments_;
            accumulated_comments_.clear();
        }

        previous_ = dialog_associate_key_[line_index_];
        line_index_ += current_data_.contents_.size();
        output_[section_][previous_] = current_data_.to_map();
        current_data_.comments_.clear();
        }
        else {
            terminated("Invalid code or content!");
        }

        // Move to the next line
        ++line_index_;
    }
}
*/
